src/lightning/data/streaming/combined.py

```diff
@@ -43,53 +49,120 @@ def __init__(
             self._weights = [w / sum(weights) for w in weights]
 
         self._iterator: Optional[_CombinedDatasetIterator] = None
+        self._use_streaming_dataloader = False
+        self._num_samples_yielded = None
+        self._current_epoch = 0
+
+    def set_epoch(self, current_epoch: int) -> None:
+        self._current_epoch = current_epoch
+        for dataset in self._datasets:
+            dataset.set_epoch(current_epoch)
+
+    def _check_datasets(self, datasets: List[StreamingDataset]) -> None:
+        if any(not isinstance(d, StreamingDataset) for d in datasets):
+            raise RuntimeError("The provided datasets should be instances of the StreamingDataset.")
+
+    def _set_use_streaming_dataloader(self, use_streaming_dataloader: bool) -> None:
+        # Used to prevent returning num_samples_yielded when using PyTorch DataLoader
+        self._use_streaming_dataloader = use_streaming_dataloader
 
     def __len__(self) -> int:
         assert self._weights
         return int(min([1 / w * len(d) for w, d in zip(self._weights, self._datasets) if w > 0]))
 
     def __iter__(self) -> Iterator[Any]:
         assert self._weights
-        self._iterator = _CombinedDatasetIterator(self._datasets, self._seed, self._weights)
+
+        worker_end = _WorkerEnv.detect()
+
+        num_samples_yielded = None
+
+        if self._num_samples_yielded and worker_end.rank in self._num_samples_yielded:
+            num_samples_yielded = self._num_samples_yielded[worker_end.rank]
+
+        self._iterator = _CombinedDatasetIterator(
+            self._datasets,
+            self._seed,
+            self._weights,
+            self._use_streaming_dataloader,
+            num_samples_yielded,
+        )
         return self._iterator
 
-    def state_dict(self, num_workers: int, batch_size: int) -> Dict[str, Any]:
+    def state_dict(
+        self, num_workers: int, batch_size: int, num_samples_yielded: Optional[List[int]] = None
+    ) -> Dict[str, Any]:
         if self._iterator is None:
-            return {}
+            if num_samples_yielded is None:
+                return {}
```

Comment:
Is it possible to call `.load_state_dict({})`?  I don't think so, because in `load_state_dict` you have this assertion: `len(state_dict["dataset"]) != len(self._datasets)`. So maybe we shouldn't return {} here, but rather a dict with all the necessary keys representing no dataset iterations. Or alternatively, implement load_state_dict differently to handle empty dicts, but that could also be error prone so maybe not. Wdyt?

Reply:
Good question. Let's accept empty state but return.